designs/2023-rule-options-defaults/README.md

```diff
@@ -0,0 +1,295 @@
+  - Drawback: `any` is generally considered a bad practice for TypeScript users as it provides no type descriptions or safety checks.
+- [`@typescript-eslint/utils` > `RuleCreator`](https://typescript-eslint.io/developers/custom-rules#rulecreator) (the recommended TypeScript-enabled rule development utility) declares a [generic `RuleContext<TMessageIds, TOptions>` type with an `options: TOptions` property](https://github.com/typescript-eslint/typescript-eslint/blob/72149e1fe8f57fd217dfc59295a51df68187aad4/packages/utils/src/ts-eslint/Rule.ts#L175).
+  - Drawback: `RuleContext` already doesn't factor in default options to the type system (https://github.com/typescript-eslint/typescript-eslint/issues/5439)
+
+Resolving the type system type resulting from a fixed JSON schema -with or without `default`- and a user-provided object can be complex.
+It is out of scope for this RFC to investigate heuristics for that problem.
+
+## Documentation
+
+We'll want to augment the following docs generators in:
+
+- The core ESLint docs: automating and therefore standardizing how rule docs pages phrase default values
+- typescript-eslint's docs: the same as the core ESLint docs
+- [`eslint-doc-generator`](https://github.com/bmish/eslint-doc-generator)
+
+We'll also want to mention this in the ESLint core custom rule documentation.
+
+## Drawbacks
+
+- This increases both conceptual and implementation complexity around rule options
+- As a future breaking change, this may break user rules - especially private rules we don't have visibility to
+
+## Backwards Compatibility Analysis
+
+The new `useRecursiveOptionDefaults` option being opt-in means this would not be a breaking change in the current version of ESLint.
+This RFC's intent is for that option to gradually be removed over time:
+
+1. ESLint v8: opt-in (default to `false`)
+2. ESLint v9: opt-out (default to `true`)
+3. ESLint v10: remove both `useRecursiveOptionDefaults` and `context.optionsRaw`
+
+Changing `useRecursiveOptionDefaults` to opt-out would be a breaking change as it impacts what values are passed to the `create` method of rules.
+That breaking change will impact rules which rely on the exact length or existing of properties in `context.options`.
+Rules that wish to preserve the legacy options behavior can switch to referring to the new `context.optionsRaw` property.
+
+In the draft PR containing this change, only 3 test files -rule tests for `indent`, `max-len`, and `no-multiple-empty-lines`- failed during `npm run test`.
+All were fixed with a direct find-and-replace from `context.options` to `context.optionsRaw` in those rules' implementations.
+
+## Alternatives
```

Comment:
It seems like another alternative that is dramatically simpler than anything mentioned in this RFC is to just allow rules to specify their defaults in JavaScript. As an example, maybe `semi` provides a default like this:

```js
meta: {
    defaultOptions: ["always"]
}
```

We'd just need to be sure to validate the default options if they're used.

To me, this has several benefits:

1. Easy to understand and change the defaults, especially for complex schemas.
2. We aren't tied into any implementation details for this from Ajv.
3. Easy to add to existing rules without needing to dig into the complexities of JSON Schema. 

We'd need to perform the merging inside of ESLint, but we already have the logic to merge rule configs, so that doesn't seem like a big deal.

Is there a reason this wouldn't work?

Reply:
This is roughly what [`@typescript-eslint/utils` > `RuleCreator`](https://typescript-eslint.io/developers/custom-rules#rulecreator) does (from `### Prior Art`). My main concern is twofold:

- A _deep_ merge is logically more complex to think about than a _shallow_ one
- This now results in _two_ places to define options: their `meta.schema` for most info, and `meta.defaultOptions` for their defaults

Also relevant: @bradzacher's https://github.com/eslint/rfcs/pull/113#issuecomment-1728496908.

It's not clear to me where to go from here. I personally am happy with either approach. If it's helpful, here's my understanding of the pros & cons:

<table>
<thead>
<tr>
<th>Approach</th>
<th>Advantages</th>
<th>Disadvantages</th>
</tr>
</thead>
<tbody>
<tr>
<th><code>meta.defaultOptions</code></th>
<td>
<ul>
<li>Easier to change and understand the defaults</li>
<li>Not tied to Ajv implementation / details</li>
</ul>
</td>
<td>
<ul>
<li>Two locations for option descriptions</li>
</ul>
</td>
</tr>
<tr>
<th>Recursive Schema Defaults</th>
<td>
<ul>
<li>One idiomatic location for option descriptions</li>
<li>Aligns closer to common Ajv paradigms</li>
</ul>
</td>
<td>
<ul>
<li>More difficult to statically type</li>
<li>Complex schemas can be ambiguous</li>
</ul>
</td>
</tr>
</tbody>
</table>

Am I missing anything? And, what happens now? Should we have a real-time discussion?

Reply:
Yes, that's helpful. Can you add these discussion points into the RFC for safekeeping?

To me, the complexity I'm most concerned about is the developer experience for people writing rules. Part of why ESLint gained popularity is because people could easily figure out how to write their own rules, and we've continued to prioritize developer experience ever since.

To me (again!), piggybacking on JSON Schema is a much more complex developer experience vs. having people specify their default options in the same format that they would in `eslint.config.js`. I'm willing to trade off complexity in the core for simplicity in the rule developer experience.

Reply:
As a data-point for deep merge vs shallow merge.

We've been using a deep merge strategy for default options as far back as [December 2018](https://github.com/bradzacher/eslint-plugin-typescript/commit/1acefaa49ee0caddfa52dfdcb71023dd9865289c), but it became the standard across our plugin rules in [February 2019](https://github.com/typescript-eslint/typescript-eslint/commit/61c60dc047da680b8cc74943c52c33562942c95a) and it became the standard across "plugins written in TS" in [May 2019](https://github.com/typescript-eslint/typescript-eslint/commit/a7a03ce39f82fac47e9ef1ef052ee1366eeb8c8a).

So for plugins written in TS deep merge has been the standard for over 4 years.

I definitely understand that not *all* usecases want deep merges - but we have evidence that it does work well for most usecases!

Reply:
Added these points to the RFC and retargeted it to suggest `meta.defaultOptions`. 👍